server/routes/snapshotListener.ts

```diff
@@ -0,0 +1,175 @@
+import { NotificationCategories } from '../../shared/types';
+import { NotificationInstance } from '../models/notification'
+
+/* NOTES FOR QA
+
+To spoof a POST request to the listener:
+Send POST to http://localhost:8080/api/snapHook with the following as the raw JSON body:
+
+  {
+    "id":"proposal/QmZ21uS8tVucpaNq2LZCbZUmHhYYXunC1ZS2gPDNWwPWD9",
+    "event":"proposal/created",   <- three other options: proposal/start, proposal/end, proposal/deleted
+    "space":"yam.eth",           <- snapshot space name goes here
+    "expire":1620947058
+  }
+
+Only chain linked to more than one snapshot space:
+    - 'hummingbot-foundation' => [ 'hbot.eth', 'hbot-ip.eth', 'hbot-prp.eth' ]
+No issues with it so far, but could be useful for testing edge cases
+
+Chains with non-standard formatting in the DB:
+    - 'exodia' => [ 'https://snapshot.org/#/exodiadao.eth' ]
+    - 'element-finance' => [ 'https://snapshot.org/#/elfi.eth' ]
+These won't work unless snapshot sends us the exact url abovce in the POST request, which I
+dont think they will do. I assume this is the user error entering the entire url rather than
+just the space
+  Ex: a POST with "space":"elfi.eth" won't work, only "space":"https://snapshot.org/#/elfi.eth" will work
+
+*/
+
+type SnapshotListenerReq = {
+  id: string;
+  event: string;
+  space: string;
+  expire: number;
+};
+
+type SnapshotListenerResp = string;
+
+const snapshotListener = async (
+  models: DB,
+  cache: SnapshotSpaceCache,
+  req: TypedRequestBody<SnapshotListenerReq>,
+  res: TypedResponse<SnapshotListenerResp>,
+) => {
+
+  // Check cache to see if any chains are subscribed to this space
+  // returns an array of the chains that subscribe to that space or an empty array
+  const chainsToNotify = await cache.checkChainsToNotify(req.body.space);
+
+  // If the array is empty, log that the listener was pinged for a invalid space
+  if (!chainsToNotify.length) {
+    mixpanelTrack({
+      event: MixpanelSnapshotEvents.SNAPSHOT_INVALID_SPACE,
+      isCustomDomain: false,
+      space: req.body.space
+    });
+    return success(res, "Snapshot POST Recieved, Space not present");
+  }
+
+  // DEPRECATED: Build array of notifications to emit from array of chains to notify
+  // const notificationObjects = chainsToNotify.map((chain:string): NotificationInstance => {
+  //   const notification = models.Notification.build({
+  //     notification_data: JSON.stringify({
+  //       created_at: new Date(),
+  //       chain_id: chain,
+  //       snapshotEventType: req.body.event,
+  //     }),
+  //     category_id: NotificationCategories.NewSnapshot,
+  //     chain_id: chain
+  //   });
+  //   return notification;
+  // });
+
+  // Temporarily emit all notifications in one step by iteratiting over each chain to notify
+  // calling the DB in a loop is bad practice, but in this case chainsToNotify
+  // will never have more than one element since no two space are subscribed to the same chain
+  // IMPORTANT TODO:
```

Comment:
TODO still pending?

Reply:
Yeah this is something either Zak or I will do this week